azure/services/managedclusters/managedclusters.go

```diff
@@ -183,23 +183,64 @@ func (s *Service) Reconcile(ctx context.Context, spec interface{}) error {
 			Type:         containerservice.VirtualMachineScaleSets,
 			VnetSubnetID: &managedClusterSpec.VnetSubnetID,
 		}
-		*properties.AgentPoolProfiles = append(*properties.AgentPoolProfiles, profile)
+		*managedCluster.AgentPoolProfiles = append(*managedCluster.AgentPoolProfiles, profile)
 	}
 
 	existingMC, err := s.Client.Get(ctx, managedClusterSpec.ResourceGroupName, managedClusterSpec.Name)
 	if err != nil && !azure.ResourceNotFound(err) {
 		return errors.Wrapf(err, "failed to get existing managed cluster")
-	} else if !azure.ResourceNotFound(err) {
+	}
+
+	isCreate := azure.ResourceNotFound(err)
+	if isCreate {
+		err = s.Client.CreateOrUpdate(ctx, managedClusterSpec.ResourceGroupName, managedClusterSpec.Name, managedCluster)
+		if err != nil {
+			return fmt.Errorf("failed to create managed cluster, %w", err)
+		}
+	} else {
 		ps := *existingMC.ManagedClusterProperties.ProvisioningState
 		if ps != "Canceled" && ps != "Failed" && ps != "Succeeded" {
-			klog.V(2).Infof("Unable to update existing managed cluster in non terminal state.  Managed cluster must be in one of the following provisioning states: canceled, failed, or succeeded")
+			s.Scope.V(2).Info("Unable to update existing managed cluster in non terminal state.  Managed cluster must be in one of the following provisioning states: canceled, failed, or succeeded")
 			return nil
 		}
-	}
 
-	err = s.Client.CreateOrUpdate(ctx, managedClusterSpec.ResourceGroupName, managedClusterSpec.Name, properties)
-	if err != nil {
-		return fmt.Errorf("failed to create or update managed cluster, %w", err)
+		// Normalize properties for the existing managed cluster, so that we
+		// check only those fields that were specified in the initial
+		// CreateOrUpdate request. Without comparing to normalized properties,
+		// we would always get a difference in desired and existing, which
+		// would result in sending unnecessary Azure API requests.
+		existingMCPropertiesNormalized := &containerservice.ManagedClusterProperties{
+			NodeResourceGroup:       existingMC.ManagedClusterProperties.NodeResourceGroup,
+			DNSPrefix:               existingMC.ManagedClusterProperties.DNSPrefix,
+			KubernetesVersion:       existingMC.ManagedClusterProperties.KubernetesVersion,
+			LinuxProfile:            existingMC.ManagedClusterProperties.LinuxProfile,
```

Comment:
don't diff the ssh keys, as I found out recently they are unfortunately not updateable -_-

Reply:
> I found out recently they are unfortunately not updatable -_-

+1 :)

Yeah, makes sense not to check them. Probably some other fields here are also immutable and can also be skipped in the check, so I'll check that out.